src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.ValueTaskSource.cs

```diff
@@ -1,235 +1,76 @@
-                                catch (ObjectDisposedException) { } // in case the SafeHandle is (erroneously) closed concurrently
-                            }
-                        }, this);
-                    }
-                    catch (OutOfMemoryException)
-                    {
-                        // Just in case trying to register OOMs, we ignore it in order to
-                        // protect the higher-level calling code that would proceed to unpin
-                        // memory that might be actively used by an in-flight async operation.
-                    }
-                }
-            }
-
-            internal void ReleaseResources()
-            {
-                // Ensure that any cancellation callback has either completed or will never run, so that
-                // we don't try to access an overlapped for this operation after it's already been freed.
-                _cancellationRegistration.Dispose();
-
-                // Unpin any pinned buffer.
-                _memoryHandle.Dispose();
-
-                // Free the overlapped.
-                if (_overlapped != null)
-                {
-                    _pipeStream._threadPoolBinding!.FreeNativeOverlapped(_overlapped);
-                    _overlapped = null;
-                }
-            }
-
-            // After calling Read/WriteFile to start the asynchronous operation, the caller may configure cancellation,
-            // and only after that should we allow for completing the operation, as completion needs to factor in work
-            // done by that cancellation registration, e.g. unregistering.  As such, we use _result to both track who's
-            // responsible for calling Complete and for passing the necessary data between parties.
-
-            /// <summary>Invoked when the async operation finished being scheduled.</summary>
-            internal void FinishedScheduling()
-            {
-                // Set the value to 1.  If it was already non-0, then the asynchronous operation already completed but
-                // didn't call Complete, so we call Complete here.  The read result value is the data (packed) necessary
-                // to make the call.
-                ulong result = Interlocked.Exchange(ref _result, 1);
-                if (result != 0)
+                switch (errorCode)
                 {
-                    Complete(errorCode: (uint)result, numBytes: (uint)(result >> 32) & 0x7FFFFFFF);
+                    // One side has closed its handle or server disconnected.
+                    // Set the state to Broken and do some cleanup work
+                    case Interop.Errors.ERROR_BROKEN_PIPE:
+                    case Interop.Errors.ERROR_PIPE_NOT_CONNECTED:
+                    case Interop.Errors.ERROR_NO_DATA:
+                    // The handle was closed
+                    case Interop.Errors.ERROR_INVALID_HANDLE:
+                        _pipeStream.State = PipeState.Broken;
```

Comment:
the need of updating this state is one of the reasons why relaxing `RandomAccess` requirements to support non-seekable files was no the way to go to share the code between `FileStream` and `PipeStream`. We could ofc just `await` the operations and call last error, but it would not be performant.

Reply:
> the need of updating this state is one of the reasons why relaxing RandomAccess requirements to support non-seekable files was no the way to go to share the code between FileStream and PipeStream

Can you elaborate?  I'd really like to get to a place where PipeStream doesn't need its own implementation and can just use RandomAccess.

Reply:
For some errors, `PipeStream` modifies it's state before throwing an exception or returning result. This state is publicly exposed (`IsMessageComplete` for example).

https://github.com/dotnet/runtime/blob/5a89bbe3a7a512e5304dfbc502c8553c7b5ebb38/src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.ValueTaskSource.cs#L165-L185

In case of `IsMessageComplete` we would need to have sth like this:

```cs
public class PipeStream
{
    public ValueTask<int> ReadAsync(Buffer, CancellationToken)
    {
        int result = await RandomAccess.ReadAsync(!Handle, Buffer, CancellationToken);

        if (LastError() == ERROR_MORE_DATA)
        {
            UpdateMessageCompletion(messageCompletion)
        }
        
        return result;
    }
}
```

Reply:
Another problem is that `SafePipeStream` binds a handle to the thread pool and currently there are no public APIs that would allow for creating `SafeFileHandle` out of `SafePipeHandle`.

Most likely both `SafePipeHandle` and `SafeFileHandle` would need to expose `ThreadPoolBoundHandle` properties (Windows-only, low level, easy to mess up things):

https://github.com/dotnet/runtime/blob/5a89bbe3a7a512e5304dfbc502c8553c7b5ebb38/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs#L31

https://github.com/dotnet/runtime/blob/5a89bbe3a7a512e5304dfbc502c8553c7b5ebb38/src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Windows.cs#L14-L17

and add ctors that accept a handle and `ThreadPoolBoundHandle`